File: experiment1/logiclayer/src/main/scala/se/alejo720/experiment1rq1/logiclayer/LogicLayer.scala
package se.alejo720.experiment1rq1.logiclayer
import java.time.{Duration, Instant}

import se.alejo720.experiment1rq1.interfaces.{DataLayerFactory, DataLayerType}

import scala.collection.mutable.ArrayBuffer

class Movie() {
    val score = 10.0f
}

class LogicLayer {
    val starts = Instant.now()
    val factory: DataLayerFactory = new DataLayerFactory()
    val ends = Instant.now()
    println("Logic-layer " + Duration.between(starts, ends).toMillis + " ms")
    //val ends = Instant.now()
    //println("Init DataLayerFactory " + Duration.between(starts, ends).toMillis + " ms")
    var datalayer: DataLayerType = null

    createDatalayer











    private def createDatalayer: Unit = {
        datalayer = factory.create()
    }










/*    def oldcalculateAverageScore(genre: String): Float = {
        var scores: Array[Float] = datalayer.getMovieScores(genre)
        val average: Float = scores.sum./(scores.length)
        average
    }*/

    def calculateAverageScoreKotlinVersion(genre: String): Float = {
        var scores: Array[Float] = datalayer.getMovieScores(genre)

        var sum = 0.0f
        for (score <- scores) {
            sum+=score
        }

        sum./(scores.length)
    }






    def average(list: Iterable[Float]): Float = {
        list.sum./(list.size)
    }

    def calculateAverageScore(genre: String): Float = {
       average(datalayer.getMovieScores(genre))
    }





    def tmpcalculateAverageScore(genre: String): Float = {
        val starts = Instant.now()

        val movieScores: Array[Float] = datalayer.getMovieScores(genre)

        val ends = Instant.now()
        println("Logic-layer calc-score: " + Duration.between(starts, ends).toMillis + " ms")

        average(movieScores)
    }

    /*def getGenres: Array[String] = {
        datalayer.getGenres()
    }*/

    def getGenres: Array[String] = {
        var scalaGenres = ArrayBuffer[String]()
        val genres = datalayer.getGenres()
        for(genre <- genres) {
            scalaGenres.+=(genre)
        }
        scalaGenres.toArray
    }

/*
    def testInteger(): Int = {
        5
    }
*/

    // Scala code
    def testNullable(): Movie ={
        null
    }

    /*def average(list: Iterable[Float]): Float = {
        list.sum./(list.size)
    }

    def calculateAverageScore(genre: String): Float = {
        val movies: List[Movie] = datalayer.getMovieScores(genre)
        average(movies.map(movie => movie.score))
    }*/
}

/*object LogicLayerMain {
  def main(args: Array[String]): Unit = {
    println("Hello World!")
    sayHello("Scala")

    calculateAverageScore("Action")
  }

  def sayHello(name: String): Unit = {
    println("Hello " + name)
  }

  private def createDatalayer(): DataLayerType = {
    val factory = new DataLayerFactory()
    val obj = factory.create()
    obj
  }

  def calculateAverageScore(genre: String): Float = {
    var datalayer: DataLayerType = createDatalayer()
    var movies: Array[Float] = datalayer.getMovieScores(genre)

    var sum: Float = 0.0f
    var score: Float = 0.0f
    for (movie <- movies) {
      println(movie)
      score = movie.toFloat
      sum = sum.+(score)
    }
    val average: Float = sum./(movies.length)
    println("SUM: " + sum)
    println("Average: " + average)
    return 0.0f
  }
}*/

File: experiment2/logiclayer/src/main/kotlin/se/alejo720/experiment1partb/logiclayer/LogicLayer.kt
package se.alejo720.experiment1partb.logiclayer
import se.alejo720.experiment1partb.datalayer.*
import java.time.Duration
import java.time.Instant

/*
C:\Users\AlexanderJ\Google Drive\Universitet\Exjobb\intellijidea-projects\experiment1partB\datalayer\build\libs\datalayer-1.0-SNAPSHOT.jar
* */
/*fun main(args: Array<String>) {
    println("(LogicLayer) Hello world from Kotlin!")
    LogicLayer()
}*/

class LogicLayer {

    val datalayer = DataLayer()

    init {
        val starts = Instant.now()
        println("(LogicLayer) Initializing")
        println("(LogicLayer) Initialized")
        val ends = Instant.now()
        println("Logic-layer: " + Duration.between(starts, ends).toMillis() + " ms")
    }

    fun calculateAverageScore(genre: String): Float {
        val starts = Instant.now()

        val movies: scala.collection.immutable.List<Movie> = datalayer.getMoviesByGenre(genre)

        var sum = 0.0f
        for (movie in movies) {
            sum+=movie.avgScore()
        }

        val average = sum / movies.size()

        val ends = Instant.now()
        println("Logic-layer calc-score: " + Duration.between(starts, ends).toMillis() + " ms")
        return average
    }

    fun getGenres(): List<String> {
        val ktlGenres = mutableListOf<String>()
        val genres = datalayer.getGenres()
        for(genre in genres) {
            ktlGenres.add(genre)
        }
        return ktlGenres
    }
}
